### DISCUSSION TITLE
How to ensure the previous partition of a job has succeeded before running the next partition

### DISCUSSION THREAD

**Discussion by yeachan153** [dagster-io/dagster#10264] [Q&A] [answered]
Hi all. We have a partitioned job which we'd like to avoid scheduling if the previous partition failed, and consequently resume scheduling subsequent partitions after the failed partition succeeds (e.g. after fixing and running a backfill for that partition). For example, if the scheduled daily partition of 2022-01-01 fails, we'd like to avoid scheduling 2022-01-02 the next day until 2022-01-01 is fixed. Only then, we'd like the scheduler to schedule 2022-01-02 and any other future missed partitions. That means that we'd need to:

- Avoid enqueuing the next partition to be scheduled if the previous partition failed
- Once the previous partition is successful again, enqueue the partitions that were temporarily not run
- Control of how many job runs that can run at the same time (doable with tags)

What would be the suggested approach to implemented this? Thanks for your help!

**Comment by sryza** [comment]
Hey @yeachan153 - do you need this because some asset partitions depend on prior partitions of upstream assets? We're currently working on adding support for partitions to the asset reconciliation sensor. Once we've done that, you'll be able to define partitions mappings that capture those dependencies on prior partitions, and then the sensor will automatically handle submitting runs when the prior partitions are available.

In the mean time, here's a workaround that should work. I haven't had a chance to test it, so there might be off-by-one errors:

```python
from bisect import bisect_left
from typing import Optional
from dagster import RunsFilter, sensor, PartitionsDefinition


def get_next_partition_key(
    partitions_def: PartitionsDefinition, partition_key: str
) -> Optional[str]:
    partition_keys = partitions_def.get_partition_keys()
    index = bisect_left(partition_keys, partition_key)
    if index >= 0 and index < len(partitions_def) - 1:
        return partition_keys[index + 1]
    else:
        return None


partitioned_job = ...


@sensor(job=partitioned_job)
def sensor(context):
    if context.cursor is None:
        partition_key = partitioned_job.partitions_def.get_partition_keys()[0]
        context.update_cursor(partition_key)
        return partitioned_job.run_request_for_partition(partition_key)
    else:
        prev_partition_key = context.cursor

        prev_partition_filled = (
            len(
                context.instance.get_runs(
                    filters=RunsFilter(tags={"dagster/partition": prev_partition_key}), limit=1
                )
            )
            > 0
        )

        if prev_partition_filled:
            next_partition_key = get_next_partition_key(prev_partition_key)
            if next_partition_key:
                context.update_cursor(next_partition_key)
                return partitioned_job.run_request_for_partition(next_partition_key)

    return None
```

**Comment by yeachan153** [comment]
Thanks @sryza! So would the sensor approach be the recommended way to go, rather than changing something in the behaviour of the scheduler_daemon/run-coordinator/launcher?

For some context, we expect to have ~500 partitioned jobs. So if we go the sensor approach we'd have ~500 sensors and basically use the sensors to kick of new job-runs in place of a schedule. We haven't done this before, so not sure if there are downsides to doing this vs using the scheduling.

> do you need this because some asset partitions depend on prior partitions of upstream assets? 

Yeah we do. For example, we get around ETL'ing some very large tables by doing a delta extract based on a column like `update_timestamp` everyday, instead of ETL'ing a billion rows daily. We then have a downstream OP in that job that basically aggregates all these deltas everyday to re-create the latest state of the table. This basically needs all the data from previous partitions for an accurate re-creation, or "snapshot" of the database at it's latest/current state.

The above doesn't apply to all our jobs, but we wanted to start by enforcing this behaviour since it also helps us check which partitions failed by inspecting the Dagit UI. At the moment, we get to see the status of the last X runs of a job in the overview page of jobs. If partitions just kept running, it would be harder of us to see based which partitions for which jobs failed.

**Reply by sryza** [reply, marked as answer]
I should update my answer - about a month ago we added direct support for self-partition dependencies. Here's an example of how to write one: https://github.com/dagster-io/dagster/discussions/11829.

If you use `build_asset_reconciliation_sensor`, you can target it at a bunch of assets at once and it will respect the self-partition dependencies and avoid materializing partitions before earlier partitions are materialized.

**Reply by yeachan153** [reply]
Thank you - I will reply there!
